### DISCUSSION TITLE
Rev. 23.3.0R33 - Pascal T3.2-23.0 incomplete?

### DISCUSSION THREAD

**Discussion by SirNibble** [prirun/p50em#17] [General]
Ran TRANSLATORS>TRANSLATORS.INSTALL.CPL - reboot

Ran PASCAL>PASCAL.INSTALL.CPL - reboot

Ran PASCAL>INSTALL_TEST*>PASCAL.INSTALL_TEST.CPL and get ..

```
<USERS>SMAN>cpl pascal.install_test.cpl
    pascal pascal>install_test*>pascal.install_test.pascal
[PASCAL Rev. T3.2-23.0 Copyright (c) 1992, Prime Computer, Inc.]
[Serial #ABC1-DEFG23-HI3J (PRIME COMPUTER)]
0000 ERRORS [PASCAL Rev. T3.0-23.0]
bind
         lo pascal.install_test
         li paslib
         li
         file install
         q
[BIND Rev. T3.2-23.0 Copyright (c) 1992, Prime Computer, Inc.]
[Serial #ABC1-DEFG23-HI3J (PRIME COMPUTER)]
:          lo pascal.install_test
:          li paslib
BIND COMPLETE
:          li
:          file install
The file is in use.  The old file is now called INSTALL.RP1.
    r install

Error: condition "ACCESS_VIOLATION$" raised at 3034(3)/44644.
(Referencing 4235(3)/2371).
```

PASCAL>PASCAL.INSTALL.CPL doesn't appear to copy anything into SYSTEM etc

Is this normal?  

PASCAL_LIBRARY.RUN is in LIBRARIES* and referenced by LSR

Everything else (CC, CBL, FTN and F77) pass their INSTALL_TEST*>

**Comment by drboone** [comment]
IIRC the rev23 PASCAL compiler is an EPF, not a statically linked executable, and therefore won't have anything in SYSTEM.

I'll have a whirl with the kit and see if I can figure out what's going on.

**Comment by drboone** [comment]
I can reproduce this.  Frustratingly, it breaks differently in the debugger.

**Comment by SirNibble** [comment]
A quick test on the console of the example 23.4.Y2K.R1 machine gives 


```
Error: condition "ACCESS_VIOLATION$" raised at 3023(3)/44644.
(Referencing 4172(3)/2371).

```

**Comment by drboone** [comment]
The actual fault happens in the PASCAL library EPF, which is mapped into segment 3023 in your second example, 3034 in your first.  The 3xxx range of segments seems to be where registered EPFs get mapped.  I unregistered the library, and it got mapped into 4703 instead.

The routine in that library where the issue occurs is P$AIIN, which seems to be involved with input file processing, perhaps setting up a default INPUT as declared in the test program's header.

So far, I haven't made any progress in figuring out what goes wrong.

Under the debugger, I get the PASSED message just before a pointer fault.  Heisenbugs.  Ugh.

**Comment by SirNibble** [comment]
Don't know if this helps or hinders (or is neither here'n'there :) ) but Static Mode still works.

```
_program hello(input,output);

begin
  writeln('Hello World');
end._

pascal hello.pascal
[PASCAL Rev. T3.2-23.0 Copyright (c) 1992, Prime Computer, Inc.]
[Serial #ABC1-DEFG23-HI3J (PRIME COMPUTER)]
0000 ERRORS [PASCAL Rev. T3.0-23.0]
<USERS>SMAN>bind
[BIND Rev. T3.2-23.0 Copyright (c) 1992, Prime Computer, Inc.]
[Serial #ABC1-DEFG23-HI3J (PRIME COMPUTER)]
: lo hello
: li paslib
BIND COMPLETE
: file
<USERS>SMAN>r hello

Error: condition "ACCESS_VIOLATION$" raised at 3034(3)/44644.
(Referencing 4221(3)/1167).
066 ER. seg -load
[SEG Rev. T3.2-23.0 Copyright (c) 1992, Prime Computer, Inc.]
[Serial #ABC1-DEFG23-HI3J (PRIME COMPUTER)]
$ lo hello
$ li paslib
LOAD COMPLETE
$ file
COMMAND ERROR
$ save
$ quit
067 ER. seg hello
Hello World

```

**Comment by drboone** [comment]
In staring at traces'n'such I've discovered that there are some dynamic links to pascal library routines which get mapped into the procedure segment.  The procedure segment is not writable in a straightforward bound executable.  The first thing the library tries to do is modify one of those dynamic links.  I'm guessing this is some kind of homegrown link snapping mechanism, though why they didn't just use the standard PRIMOS one eludes me.  Anyway, since that's an attempted write to a non-writable memory location, ACCESS_VIOLATION$.  This explains why the SEG-loaded version works, and probably why it doesn't explode in that way under the debugger.  It's going to take some digging to work out whether this bug belongs to the compiler, bind, or a collaboration.

**Comment by drboone** [comment]
Further digging has taught me that P$AINP and P$AOUT are _common_ areas.  I almost accidentally discovered that if the link sequence looks like this::
```
bind
lo pascal.install_test.bin
rdc
li paslib
file install.run
```
then the resulting runfile works.

**Comment by SirNibble** [comment]
Yep! That works. 

Cheers!

**Comment by SirNibble** [comment]
I can get back to my "can I invoke a logout.cpl program on LOGOUT" project now.

Is there manual for this revision? I'm guessing the Rev19 one are probably OK, but best to check.

**Comment by drboone** [comment]
Unfortunately, the latest Pascal manual we have is 19.2.

The matching manual for the rev23 compiler would be something like DOC4303-4LA, which I think would say rev22 or rev23.  Maybe we'll get lucky some day.

**Comment by SirNibble** [comment]
It was an act of purest optimism to ask in the first place! :)

**Comment by drboone** [comment]
With Jim's help I think I've fixed the PASCAL problem in the latest commit.  The MODULA and PL1 issues I've seen didn't respond to this fix, so more work required on those.

**Comment by drboone** [comment]
Actually, after re-compiling, the PL1 hello-world program _does_ work now.

**Comment by SirNibble** [comment]
Re-made the emulator on Raspbian Buster and the PASCAL>INSTALL_TEST*>PASCAL.INSTALL_TEST.CPL now completes as expected.

Was it an easy fix?
